# Data Preprocessing

For our mini-challenge, we will be utilizing the Cleantech Media Dataset, which serves as an invaluable resource for businesses, researchers, and students interested in the latest developments in Natural Language Processing and Large Language Models within the realm of cleantech and sustainability. In an industry that is constantly evolving, having access to timely and accurate information is crucial. This dataset is specifically designed to address those needs.

This dataset is accessible on Kaggle and is credited to [Janna Lipenkova](https://www.kaggle.com/datasets/jannalipenkova/cleantech-media-dataset).

## Imports

In [378]:
import os
import pandas as pd
import numpy as np

import re
import nltk
from nltk.corpus import stopwords
from bs4 import BeautifulSoup
from nltk.stem import PorterStemmer, WordNetLemmatizer

# Ensure you have the NLTK stopwords
nltk.download('stopwords')
stop_words = set(stopwords.words('english'))

# Initialize stemmer and lemmatizer
stemmer = PorterStemmer()
lemmatizer = WordNetLemmatizer()

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/alexanderschilling/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


## Data

### Training Data

- Comprehensive Coverage: Access a wide range of media texts on cleantech topics, from renewable energy to carbon reduction.
- Efficiency: Utilize the dataset for quick and accurate question-answering, aiding informed decision-making.
- Regular Updates: Stay current with monthly updates reflecting the latest trends in cleantech.
- Sustainability Focus: Contribute to the sustainability movement by leveraging valuable insights from the dataset.

In [379]:
data = pd.read_csv('../data/cleantech_media_dataset_v2_2024-02-23.csv')
data.head()

Unnamed: 0                                              title        date  \
0        1280  Qatar to Slash Emissions as LNG Expansion Adva...  2021-01-13   
1        1281               India Launches Its First 700 MW PHWR  2021-01-15   
2        1283              New Chapter for US-China Energy Trade  2021-01-20   
3        1284  Japan: Slow Restarts Cast Doubt on 2030 Energy...  2021-01-22   
4        1285     NYC Pension Funds to Divest Fossil Fuel Shares  2021-01-25   

  author                                            content       domain  \
0    NaN  ["Qatar Petroleum ( QP) is targeting aggressiv...  energyintel   
1    NaN  ["• Nuclear Power Corp. of India Ltd. ( NPCIL)...  energyintel   
2    NaN  ["New US President Joe Biden took office this ...  energyintel   
3    NaN  ["The slow pace of Japanese reactor restarts c...  energyintel   
4    NaN  ["Two of New York City's largest pension funds...  energyintel   

                                                 url  
0  https://www.energyintel.com/0000017b-a7dc-de4c...  
1  https://www.energyintel.com/0000017b-a7dc-de4c...  
2  https://www.energyintel.com/0000017b-a7dc-de4c...  
3  https://www.energyintel.com/0000017b-a7dc-de4c...  
4  https://www.energyintel.com/0000017b-a7dc-de4c...

### Evaluation data

- The dataset comes with a small, high-quality evaluation dataset for the "Retrieval" step in Retrieval-Augmented Generation
- In addition, a small collection of human gold-standard query-passage-answer triplets will be provided for further evaluation

In [380]:
data_eval = pd.read_csv('../data/cleantech_rag_evaluation_data_2024-09-20.csv', delimiter=";")
data_eval.head()

example_id  question_id                                           question  \
0           1            1  What is the innovation behind Leclanché's new ...   
1           2            2       What is the EU’s Green Deal Industrial Plan?   
2           3            2       What is the EU’s Green Deal Industrial Plan?   
3           4            3  What are the four focus areas of the EU's Gree...   
4           5            4  When did the cooperation between GM and Honda ...   

                                       relevant_text  \
0  Leclanché said it has developed an environment...   
1  The Green Deal Industrial Plan is a bid by the...   
2  The European counterpart to the US Inflation R...   
3  The new plan is fundamentally focused on four ...   
4  What caught our eye was a new hookup between G...   

                                              answer  \
0  Leclanché's innovation is using a water-based ...   
1  The EU’s Green Deal Industrial Plan aims to en...   
2  The EU’s Green Deal Industrial Plan aims to en...   
3  The four focus areas of the EU's Green Deal In...   
4                                          July 2013   

                                         article_url  
0  https://www.sgvoice.net/strategy/technology/23...  
1  https://www.sgvoice.net/policy/25396/eu-seeks-...  
2  https://www.pv-magazine.com/2023/02/02/europea...  
3  https://www.sgvoice.net/policy/25396/eu-seeks-...  
4  https://cleantechnica.com/2023/05/08/general-m...

## Initial Exploration

In [381]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9593 entries, 0 to 9592
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  9593 non-null   int64 
 1   title       9593 non-null   object
 2   date        9593 non-null   object
 3   author      31 non-null     object
 4   content     9593 non-null   object
 5   domain      9593 non-null   object
 6   url         9593 non-null   object
dtypes: int64(1), object(6)
memory usage: 524.7+ KB
None


In [382]:
# Check for missing values
print(data.isnull().sum())

Unnamed: 0       0
title            0
date             0
author        9562
content          0
domain           0
url              0
dtype: int64


In [383]:
print(data["domain"].unique())

['energyintel' 'energyvoice' 'eurosolar' 'indorenergy' 'cleantechnica'
 'decarbxpo' 'ecofriend' 'greenprophet' 'azocleantech' 'naturalgasintel'
 'businessgreen' 'rechargenews' 'solarquarter' 'thinkgeoenergy'
 'pv-magazine' 'solarpowerworldonline' 'pv-tech' 'solarpowerportal.co'
 'solarindustrymag']


Now we want to see what the text oin the column "content" looks like. This will give us a better idea on how the texts are structured so that we can make a good plan for the data cleaning process.

In [384]:
data["content"][100]

"['Velocys has signed two offtake agreements that cover the entire sustainable aviation fuel ( SAF) production from its Bayou Fuels plant in Mississippi for at least ten years from start up in 2026. US Southwest Airlines will take-or-pay for 219 million gallons of carbon-negative SAF — generating 575 million gallons of carbon neutral blended SAF — over 15 years. Europe’ s IAG will take 73 million gallons of carbon negative SAF — 192 million gallons of carbon-neutral SAF after blending — over 10 years. Velocys said the deal with IAG was worth $ 800 million. Both deals include a price support mechanism for the greenhouse gas credits associated with Bayou’ s SAF production, which claims a carbon reduction of 6.5 million tons over the term of the contract. The plant has negative lifecycle emissions thanks to its biogenic waste feedstock, renewable power and carbon capture and storage. “ This long-dated offtake, encompassing support for environmental credits, will provide certainty of reven

## Text Preprocessing

**Steps to clean the data**
- convert to lowercase
- some content rows contain urls -> remove them
- remove special characters (for example ", !, ?, ', `)
- replace "%" with "percent"
- "/" between letters or words should not be removed (for example "tons/yr" should become "tons per year" and not "tonsyr")
- "-" that are word connectors (for example "carbon-negative") should not be removed -> maybe a space could be made instead
- "-" that are between two empty spaces (for example "SAF — 192 million gallons") -> these should be removed
- currency symbols should be replaced by written words (for example $ turns to dollars)
- Country abbreviations can be the same as words (for example US and us) -> what can we do here?

### Lowercase

In [385]:
def to_lowercase(text):
    return text.lower()

data['content_cleaned'] = data['content'].apply(to_lowercase)

### Remove HTML tags

In [386]:
def remove_html_tags(text):
    return BeautifulSoup(text, "html.parser").get_text()

data['content_cleaned'] = data['content_cleaned'].apply(remove_html_tags)

### Remove URLs

In [387]:
def remove_urls(text):
    return re.sub(r'http\S+|www\S+|https\S+', '', text)

data['content_cleaned'] = data['content_cleaned'].apply(remove_urls)

### Remove Special Characters

In [388]:
def remove_special_characters(text):
    # Keep alphanumeric characters, certain punctuation, and hyphens/slashes while removing unwanted special characters
    # Allow periods, commas, question marks, exclamation marks, percentage symbols, and ampersands
    text = re.sub(r"[\\\'\"`]", '', text)  # Remove backslashes, single quotes, double quotes, and backticks
    text = re.sub(r'[^A-Za-z0-9\s\-\/\.\,\?\!\%&]', '', text)  # Keep allowed punctuation marks

    # Replace "/" with " per " and specific words with their full forms
    text = re.sub(r'(\w+)/(\w+)', lambda m: f"{m.group(1)} per {'year' if m.group(2) == 'yr' else m.group(2)}", text)
    
    text = re.sub(r'\s*-\s*', ' ', text)  # Replace hyphens surrounded by spaces with a space
    text = re.sub(r'(?<!\w)-|-(?!\w)', '', text)  # Remove hyphens not surrounded by words

    result = result.replace('%', ' percent')

    # remove double space
    text = re.sub(r'\s+', ' ', text).strip()
    
    return text

data['content_cleaned'] = data['content_cleaned'].apply(remove_special_characters)

In [389]:
data['content_cleaned'][0]

'qatar petroleum qp is targeting aggressive cuts in its greenhouse gas emissions as it prepares to launch phase 2 of its planned 48 million ton per year lng expansion. in its latest sustainability report published on wednesday, qp said its goals include reducing the emissions intensity of qatars lng facilities by 25% and of its upstream facilities by at least 15%. the company is also aiming to reduce gas flaring intensity across its upstream facilities by more than 75% and has raised its carbon capture and storage ambitions from 5 million tons per year to 7 million tons per year by 2027. about 2.2 million tons per year of the carbon capture goal will come from the 32 million ton per year phase 1 of the lng expansion, also known as the north field east project. a further 1.1 million tons per year will come from phase 2, known as the north field south project, which will raise qatars lng capacity by a further 16 million tons per year. qatar currently has an lng production capacity of aro

### Remove punctuation

- 0.2% would turn to 02 
- 2.2 million would turn into 22 million
-> we have to be careful how we delete the punctuations

In [390]:
def remove_periods(text):
    # This regex pattern matches a period that is NOT followed by a digit
    pattern = r'\.(?!\d)'
    # Replace matched periods with an empty string
    result = re.sub(pattern, '', text)
    # Remove all commas
    result = result.replace(',', '')
    return result

data['content_cleaned'] = data['content_cleaned'].apply(remove_periods)

### Remove stop words

In [391]:
def remove_stop_words(text):
    tokens = text.split()
    tokens = [word for word in tokens if word not in stop_words]
    return ' '.join(tokens)

data['content_cleaned'] = data['content_cleaned'].apply(remove_stop_words)

### Tokenization

In [392]:
def tokenize(text):
    return text.split()

data['content_tokenized'] = data['content_cleaned'].apply(tokenize)

### Stemming

here we will stemm the words in the content column

In [393]:
def stem_text(tokens):
    return ' '.join(stemmer.stem(word) for word in tokens)

data['content_stemmed'] = data['content_tokenized'].apply(stem_text)

### Lemmatization

here we will lemmatize the words in the content column

In [394]:
def lemmatize_text(tokens):
    return ' '.join(lemmatizer.lemmatize(word) for word in tokens)

data['content_lemmatized'] = data['content_tokenized'].apply(lemmatize_text)

### Result

In [395]:
data["content"][0]

'["Qatar Petroleum ( QP) is targeting aggressive cuts in its greenhouse gas emissions as it prepares to launch Phase 2 of its planned 48 million ton per year LNG expansion. In its latest Sustainability Report published on Wednesday, QP said its goals include `` reducing the emissions intensity of Qatar\'s LNG facilities by 25% and of its upstream facilities by at least 15%. \'\' The company is also aiming to reduce gas flaring intensity across its upstream facilities by more than 75% and has raised its carbon capture and storage ambitions from 5 million tons/yr to 7 million tons/yr by 2027. About 2.2 million tons/yr of the carbon capture goal will come from the 32 million ton/yr Phase 1 of the LNG expansion, also known as the North Field East project. A further 1.1 million tons/yr will come from Phase 2, known as the North Field South project, which will raise Qatar\'s LNG capacity by a further 16 million tons/yr. Qatar currently has an LNG production capacity of around 78 million tons

In [396]:
data['content_cleaned'][0]

'qatar petroleum qp targeting aggressive cuts greenhouse gas emissions prepares launch phase 2 planned 48 million ton per year lng expansion latest sustainability report published wednesday qp said goals include reducing emissions intensity qatars lng facilities 25% upstream facilities least 15% company also aiming reduce gas flaring intensity across upstream facilities 75% raised carbon capture storage ambitions 5 million tons per year 7 million tons per year 2027 2.2 million tons per year carbon capture goal come 32 million ton per year phase 1 lng expansion also known north field east project 1.1 million tons per year come phase 2 known north field south project raise qatars lng capacity 16 million tons per year qatar currently lng production capacity around 78 million tons per year eyeing phased expansion 126 million tons per year qp says able eliminate routine gas flaring 2030 methane emissions limited setting methane intensity target 0.2% across facilities 2025 company also plans

In [397]:
data['content_stemmed'][0]

'qatar petroleum qp target aggress cut greenhous ga emiss prepar launch phase 2 plan 48 million ton per year lng expans latest sustain report publish wednesday qp said goal includ reduc emiss intens qatar lng facil 25% upstream facil least 15% compani also aim reduc ga flare intens across upstream facil 75% rais carbon captur storag ambit 5 million ton per year 7 million ton per year 2027 2.2 million ton per year carbon captur goal come 32 million ton per year phase 1 lng expans also known north field east project 1.1 million ton per year come phase 2 known north field south project rais qatar lng capac 16 million ton per year qatar current lng product capac around 78 million ton per year eye phase expans 126 million ton per year qp say abl elimin routin ga flare 2030 methan emiss limit set methan intens target 0.2% across facil 2025 compani also plan build 1.6 gigawatt solar energi capac 2025 half come siraj solar power project next year eif jan.2220 month littl news phase 2 qatar mas

In [398]:
data['content_lemmatized'][0]

'qatar petroleum qp targeting aggressive cut greenhouse gas emission prepares launch phase 2 planned 48 million ton per year lng expansion latest sustainability report published wednesday qp said goal include reducing emission intensity qatar lng facility 25% upstream facility least 15% company also aiming reduce gas flaring intensity across upstream facility 75% raised carbon capture storage ambition 5 million ton per year 7 million ton per year 2027 2.2 million ton per year carbon capture goal come 32 million ton per year phase 1 lng expansion also known north field east project 1.1 million ton per year come phase 2 known north field south project raise qatar lng capacity 16 million ton per year qatar currently lng production capacity around 78 million ton per year eyeing phased expansion 126 million ton per year qp say able eliminate routine gas flaring 2030 methane emission limited setting methane intensity target 0.2% across facility 2025 company also plan build 1.6 gigawatts sola

### Next steps

- Text Analysis / NLP Task Preparation
- Sentiment Analysis
- Text Summarization
- Save the cleaned data
- Exploratory Data Analysis (EDA)


the next steps could include limiting the vocabulary 